### ISSUE TITLE
nixpkgs python repl isn't very terminal-friendly

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#17607]
## Issue description

I can't use standard terminal keys in the `python` from nixpkgs.

### Steps to reproduce

1. `nix-shell -p python`
2. `python`
3. Enter `5 + 5`
4. Press the up arrow.
5. Weep when you see `^[[A`
6. Press the left arrow
7. Weep when you see `^[[D`
8. Press ctrl-A
9. Weep when you see `^A`

## Technical details

I tried it on darwin nixpkgs and on NixOS:

* NixOS version: 16.03pre73316.93d8671 (Emu)
* Darwin Nixpkgs version: 16.09pre87549.8a1e7cd

**Comment by rvl** [comment]
Readline is available if you use `nix-shell -p pythonFull`.

Do you mean that the default `python` attribute should be changed, and/or documentation added?

**Comment by copumpkin** [comment]
Yeah, perhaps my issue is just using the default `python`. Note that this also affects transitive dependencies, so if I do `nix-shell -p pythonPackages.boto3` the python I get in there will also be unfriendly.

It's also kind of awkward to have the distinction, because if I do `nix-shell -p pythonPackages.boto3 pythonFull` I end up with no readline, but `nix-shell -p pythonFull pythonPackages.boto3` gives me a version with readline.

I don't know why the default attribute is the weaker version, and I don't necessarily want to replace it by `pythonFull`, but I do want the overall default experience to be good for people coming from other systems. As it stands, it seems like making readline optional adds some confusion.

**Comment by FRidh** [comment]
By default `python` is a stripped version. All packages that use `buildPythonPackage` also depend on this version. 

> It's also kind of awkward to have the distinction, because if I do nix-shell -p pythonPackages.boto3 pythonFull I end up with no readline, but nix-shell -p pythonFull pythonPackages.boto3 gives me a version with readline.

I guess the `setupHook` together with `PYTHONPATH` being evaluated from left to right is responsible for that.

> I do want the overall default experience to be good for people coming from other systems

Do we want `python` to refer to a full version of Python by default, or do we want to point it to something that is sufficient for most users and especially sufficient for most other packages that depend on Python, and taking into account the closure size?

**Comment by domenkozar** [comment]
We should include readline by default. I've been long wanting to do that, but never had the time.

**Comment by copumpkin** [comment]
@domenkozar what's the downside? All python packages gain a slightly smaller runtime closure?

**Comment by domenkozar** [comment]
I think we'll survive:

```
$ du -sh $(nix-build -A readline.all)
752K    /nix/store/q97il2whcmnmcyvr62n252f1ay5i4irq-readline-6.3p08
168K    /nix/store/mbsx71lzniqhhv4fi7al50aavqmqq4a9-readline-6.3p08-dev
888K    /nix/store/2aj8lx78bfnmhmbjvh34f0amqnfibbaq-readline-6.3p08-doc
```

**Comment by copumpkin** [comment]
😱

**Comment by FRidh** [comment]
Fixes in #19309.
